evaluate/task/task-transpile.go

```diff
@@ -0,0 +1,145 @@
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+
+	pkgerrors "github.com/pkg/errors"
+	"github.com/symflower/eval-dev-quality/evaluate/metrics"
+	"github.com/symflower/eval-dev-quality/language"
+	"github.com/symflower/eval-dev-quality/language/golang"
+	"github.com/symflower/eval-dev-quality/language/java"
+	"github.com/symflower/eval-dev-quality/log"
+	"github.com/symflower/eval-dev-quality/model"
+	evaltask "github.com/symflower/eval-dev-quality/task"
+)
+
+// TaskTranspile holds the transpilation task.
+type TaskTranspile struct{}
+
+// TaskArgumentsTranspile holds extra arguments to be used in a query prompt.
+type TaskArgumentsTranspile struct {
+	// SourceLanguage holds the source language for transpilation.
+	SourceLanguage language.Language
+	// StubFilePath holds the path for the file containing just the function signature of the language we are transpiling to.
+	StubFilePath string
+}
+
+var _ evaltask.Task = (*TaskTranspile)(nil)
+
+// Identifier returns the transpilation task identifier.
+func (t *TaskTranspile) Identifier() evaltask.Identifier {
+	return IdentifierTranspile
+}
+
+// Run transpiles code between Go and Java and runs predefined tests to check if the transpilation was successful.
+func (t *TaskTranspile) Run(ctx evaltask.Context) (repositoryAssessment map[evaltask.Identifier]metrics.Assessments, problems []error, err error) {
+	modelCapability, ok := ctx.Model.(model.CapabilityTranspile)
+	if !ok {
+		return nil, nil, pkgerrors.Wrap(evaltask.ErrTaskUnsupportedByModel, fmt.Sprintf("%q does not support %q", ctx.Model.ID(), string(t.Identifier())))
+	}
+
+	taskLogger, err := newTaskLogger(ctx, t)
+	if err != nil {
+		return nil, nil, err
+	}
+	defer func() {
+		taskLogger.finalize(problems)
+	}()
+
+	var packagePaths []string
+	files, err := os.ReadDir(ctx.Repository.DataPath())
+	if err != nil {
+		return nil, nil, pkgerrors.WithStack(err)
+	}
+	for _, file := range files {
+		if file.IsDir() && !strings.HasPrefix(file.Name(), ".") { // Ignore hidden directories.
+			packagePaths = append(packagePaths, filepath.Join(ctx.Repository.DataPath(), file.Name()))
+		}
+	}
+
+	modelAssessments := metrics.NewAssessments()
+	for _, packagePath := range packagePaths {
+		if err := ctx.Repository.Reset(ctx.Logger); err != nil {
+			ctx.Logger.Panicf("ERROR: unable to reset temporary repository path: %s", err)
+		}
+
+		var sourceLanguage language.Language
+		if _, ok := ctx.Language.(*golang.Language); ok {
+			sourceLanguage = &java.Language{}
+		} else {
+			sourceLanguage = &golang.Language{}
+		}
+
+		sourceFilePath, stubFilePath, err := t.unpackTranspilerPackage(ctx, taskLogger.Logger, sourceLanguage, packagePath)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		modelContext := model.Context{
+			Language: ctx.Language,
+
+			RepositoryPath: packagePath,
+			FilePath:       sourceFilePath,
+
+			Arguments: &TaskArgumentsTranspile{
+				SourceLanguage: sourceLanguage,
+				StubFilePath:   stubFilePath,
+			},
```

Comment:
Just so I follow this correctly... Now  `.Arguments.SourceLanguage` holds the language for `.FilePath` and `.Language` holds the language for `.Arguments.StubFilePath`? Bit confusing now that it is crossed. :thinking: !

Maybe we can rename `SourceLanguage` into `OriginLanguage` and and `StubFilePath` into `OriginFilePath` to store the "origin" implementation? And please also check the docstrings and that the LLM prompt construction uses the correct structure and wording.

Reply:
I agree with changing `SourceLanguage` to `OriginLanguage`, but using `OriginFilePath` does not sound correct because is not related to the `OriginLanguage`

Reply:
Well. I would switch where `sourceFilePath` and `stubFilePath` are stored. Then `OriginFilePath` is correct.